modules/core/keeper/msg_server.go

```diff
@@ -751,7 +751,72 @@ func (k Keeper) ChannelUpgradeInit(goCtx context.Context, msg *channeltypes.MsgC
 
 // ChannelUpgradeTry defines a rpc handler method for MsgChannelUpgradeTry.
 func (k Keeper) ChannelUpgradeTry(goCtx context.Context, msg *channeltypes.MsgChannelUpgradeTry) (*channeltypes.MsgChannelUpgradeTryResponse, error) {
-	return nil, nil
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	module, _, err := k.ChannelKeeper.LookupModuleByChannel(ctx, msg.PortId, msg.ChannelId)
+	if err != nil {
+		ctx.Logger().Error("channel upgrade try failed", "port-id", msg.PortId, "error", errorsmod.Wrap(err, "could not retrieve module from port-id"))
+		return nil, errorsmod.Wrap(err, "could not retrieve module from port-id")
+	}
+
+	cbs, ok := k.Router.GetRoute(module)
+	if !ok {
+		ctx.Logger().Error("channel upgrade try failed", "port-id", msg.PortId, "error", errorsmod.Wrapf(porttypes.ErrInvalidRoute, "route not found to module: %s", module))
+		return nil, errorsmod.Wrapf(porttypes.ErrInvalidRoute, "route not found to module: %s", module)
+	}
+
+	// TODO: this may be updated based on refactors to this step
+	proposedUpgrade, err := k.ChannelKeeper.ChanUpgradeTry(
+		ctx,
+		msg.PortId,
+		msg.ChannelId,
+		msg.ProposedUpgradeConnectionHops,
+		msg.UpgradeTimeout,
+		msg.CounterpartyProposedUpgrade,
+		msg.CounterpartyUpgradeSequence,
+		msg.ProofChannel,
+		msg.ProofUpgrade,
+		msg.ProofHeight,
+	)
+	if err != nil {
+		ctx.Logger().Error("channel upgrade try failed", "error", err)
+
+		return &channeltypes.MsgChannelUpgradeTryResponse{
+			Result: channeltypes.FAILURE,
+		}, nil
+	}
+
+	proposedVersion, err := cbs.OnChanUpgradeTry(
+		ctx,
+		msg.PortId,
+		msg.ChannelId,
+		msg.CounterpartyProposedUpgrade.Fields.Ordering,
+		msg.CounterpartyProposedUpgrade.Fields.ConnectionHops,
+		msg.CounterpartyProposedUpgrade.Fields.Version,
+	)
```

Comment:
Is passing the counterparty connection hops to the application callback on the TRY handler correct? I would expect it to be the connection hops of _this_ host, i.e. `msg.ProposedUpgradeConnectionHops` OR `proposedUpgrade.ConnectionHops`.

Should we rather use the values from the returned `proposedUpgrade` above? I think it might make more sense.

```suggestion
	proposedVersion, err := cbs.OnChanUpgradeTry(
		ctx,
		msg.PortId,
		msg.ChannelId,
		proposedUpgrade.Fields.Ordering,
		proposedUpgrade.Fields.ConnectionHops,
		proposedUpgrade.Fields.Version,
	)
```

Reply:
ah true, i think this may have been a legacy from a previous refactor of the TRY handler 🙈 too many iterations to keep track of.

Reply:
Agreed!